File: solutions/java/src/main/java/com/epi/SearchMinFirstBST.java
package com.epi;

import com.epi.BinaryTreePrototypeTemplate.BinaryTreeNode;

public class SearchMinFirstBST {
  // @include
  public static boolean searchMinFirstBST(BinaryTreeNode<Integer> minFirstBST,
                                          Integer k) {
    // First handle the base cases.
    if (minFirstBST == null || Integer.compare(minFirstBST.data, k) > 0) {
      return false;
    } else if (Integer.compare(minFirstBST.data, k) == 0) {
      return true;
    }

    // Recursively search just the right subtree if the smallest key in the
    // right subtree is greater than or equal to k.
    if (minFirstBST.right != null
        && Integer.compare(k, minFirstBST.right.data) >= 0) {
      return searchMinFirstBST(minFirstBST.right, k);
    }
    return searchMinFirstBST(minFirstBST.left, k);
  }
  // @exclude

  public static void main(String[] args) {
    // A min-first BST
    // 1
    // 2 4
    // 3 5 7
    BinaryTreeNode<Integer> tree = new BinaryTreeNode<>(1);
    tree.left = new BinaryTreeNode<>(2);
    tree.left.left = new BinaryTreeNode<>(3);
    tree.right = new BinaryTreeNode<>(4);
    tree.right.left = new BinaryTreeNode<>(5);
    tree.right.right = new BinaryTreeNode<>(7);
    assert(searchMinFirstBST(tree, 1));
    assert(searchMinFirstBST(tree, 3));
    assert(searchMinFirstBST(tree, 5));
    assert(!searchMinFirstBST(tree, 6));
  }
}

File: static/python/search_min_first_bst.py
from binary_tree_prototype import BinaryTreeNode


# @include
def search_min_first_bst(min_first_bst, k):
    # First handle the base cases.
    if not min_first_bst or min_first_bst.data > k:
        return False
    elif min_first_bst.data == k:
        return True

    # Recursively search just the right subtree if the smallest key in the
    # right subtree is greater than or equal to k.
    if min_first_bst.right and k >= min_first_bst.right.data:
        return search_min_first_bst(min_first_bst.right, k)
    return search_min_first_bst(min_first_bst.left, k)
# @exclude


def main():
    # A min-first BST
    #      1
    #    2   4
    #  3    5 7
    tree = BinaryTreeNode(1)
    tree.left = BinaryTreeNode(2)
    tree.left.left = BinaryTreeNode(3)
    tree.right = BinaryTreeNode(4)
    tree.right.left = BinaryTreeNode(5)
    tree.right.right = BinaryTreeNode(7)
    assert search_min_first_bst(tree, 1)
    assert search_min_first_bst(tree, 3)
    assert search_min_first_bst(tree, 5)
    assert not search_min_first_bst(tree, 6)


if __name__ == '__main__':
    main()
